### DISCUSSION TITLE
Improve user updates from providers

### DISCUSSION THREAD

**Discussion by nbouvrette** [nextauthjs/next-auth#9764] [Ideas]
### Description 📓

There are some 3rd party providers (like LinkedIn) that change image URLs. I noticed that after users have had accounts for a while their avatar image is no longer working.

Unless I misunderstood, the [suggested](https://github.com/nextauthjs/next-auth/discussions/1414) way to update user info is to use a `signIn` callback? This is how I managed to fix images... 

```ts
callbacks: {
    /**
     * Sometimes, social providers can update the image URL and this is the only way to update it.
     */
    signIn: async ({ user, account, profile }) => {
      if (!account || !profile) {
        return true
      }

      const { provider } = account

      let providerUser: User

      if (provider === 'linkedin') {
        providerUser = await LinkedInProvider(LinkedInProviderOptions).profile(
          profile as LinkedInProfile,
          account
        )
      } else if (provider === 'google') {
        providerUser = await GoogleProvider(GoogleProviderOptions).profile(
          profile as GoogleProfile,
          account
        )
      // You would need to do this for all the providers!!
      } else {
        return true
      }

      if (providerUser.image && user.image !== providerUser.image) {
        await changeUserImageUrl(user.id, providerUser.image)
      }
      return true
    }
}
```

While this seems to work, here are my concerns:

1) It does an extra API call to the provider to get the image
2) The is no other way as far as I know to avoid the extra API call because the `profile` data returned in the callback is "raw" and the providers do not expose their logic to extract images
3) This code can become very complex to manage if you have a lot of providers and it feels like its mostly boilerplate code that should be handled somewhere else

Suggestion/idea:

Return the parsed `user` object from the provider in the callback would make this a lot simpler

But maybe even better, having some sort of flag like `updateImage: true` in the provider config that would handle this automatically by calling the adapter's `updateUser` function when images change?

### How to reproduce ☕️

Not applicable (just use the Next-auth LinkedInProvider and you will see images break after a while).

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by osdiab** [comment]
Ran into this issue - Google provider seems to include it in the profile.picture parameter but Azure AD Provider does not, which necessitates this type of code. Would be great if these values were consistently there somehow, updating images on login is not intuitive.

**Comment by davidwinter** [comment]
Same issue. I'm including the groups that are associated to a user in AWS Cognito, and if a user is removed from a group, then need a way to update that in our database record of the user.
